pkg/doctor/checks.go

```diff
@@ -0,0 +1,243 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package doctor
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+)
+
+// CheckFunc is a function that performs a check on a log line
+type CheckFunc func(line *LogEntry, report *SimpleReport)
+
+// Selectors stores all registered selector patterns and their associated check functions
+var Selectors = make(map[string]CheckFunc)
+
+// CriticalPatterns contains compiled regex patterns for identifying critical error lines
+var criticalPatterns = []*regexp.Regexp{
+	regexp.MustCompile(`(?i)^\s*Command failed:`),
+	regexp.MustCompile(`(?i)^\s*(Error|FATAL|CRITICAL)\b`),
+	regexp.MustCompile(`(?i)^\s*Caused by:`),
+	regexp.MustCompile(`(?i)^\s*[\w.]+Error:`),
+	regexp.MustCompile(`(?i)permission denied`),
+	regexp.MustCompile(`(?i)failed`),
+	regexp.MustCompile(`(?i)exception`),
+	regexp.MustCompile(`(?i)could not connect`),
+	regexp.MustCompile(`(?i)timed out`),
+}
+
+// RegisterSelector registers a selector pattern with its associated check function
+func registerSelector(selector string, checkFunc CheckFunc) {
+	Selectors[selector] = checkFunc
+}
+
+func init() {
+	// Register all selectors
+	registerSelector("Reached PR limit - skipping PR creation", prLimitReached)
+	registerSelector("Found renovate config errors", renovateConfigErrors)
+	registerSelector("rawExec err", rawExecError)
+	registerSelector("Platform-native commit: unknown error", platformCommitError)
+}
+
+// extractUsefulError extracts the most useful parts of a potentially long error message.
+func extractUsefulError(fullMessage string, maxOutputLines int) string {
+	if fullMessage == "" {
+		return ""
+	}
+
+	lines := strings.Split(fullMessage, "\n")
+	// remove trailing empty lines
+	if lines[0] == "" {
+		lines = lines[1:]
+	}
+	if lines[len(lines)-1] == "" {
+		lines = lines[:len(lines)-1]
+	}
+
+	// If short enough, return as-is
+	if len(lines) <= maxOutputLines {
+		return strings.TrimSpace(fullMessage)
+	}
+
+	return processLongMessage(lines, maxOutputLines)
+}
+
+// processLongMessage processes a long message, keeping critical lines and context
+func processLongMessage(lines []string, maxOutputLines int) string {
+	usefulLines := []string{strings.TrimSpace(lines[0])}
+	contextBuffer := make([]string, 0, 2) // keep previous 2 lines of context
+	cutLinesCount := 0
+	omittedLines := 0
+
+	// Pattern to match lines with only symbols like ~^=
+	symbolPattern := regexp.MustCompile(`^\s*[~^=]+\s*$`)
+
+	for i, line := range lines[1:] { // skip first line, already added
+		trimmedLine := strings.TrimSpace(line)
+
+		// Skip empty lines or lines with only symbols
+		if trimmedLine == "" || symbolPattern.MatchString(trimmedLine) {
+			continue
+		}
+
+		if i == len(lines)-1 {
```

Comment:
This is never true. The max value of `i` is `len(lines) -1 -1`, because the index starts from 0.